Refactoring (v3) (#96)
fastify/process-warning: README.md

### Usage

> The module exports a 2 builder functions for creating warnings and emitting them.

Reviewer: ```suggestion
The module exports two builder functions for creating warnings. A warning has a `emit` method.
```

```js
const {
  createWarning,
  createDeprecation
} = require('process-warning')

> const warning = createWarning('FastifyWarning', 'FSTWRN001', 'Hello %s', { unlimited: true })

Reviewer 2: If we are redoing the whole interface, what is everyone's opinion of?:

```js
const warning = createWarning({
  name: 'ExampleWarning',
  code: 'EXMPWRN001',
  message: 'Hello %s',
  interpolationValues: ['world'],
  unlimited: true
})
```

Reviewer: What is interpolationValues?

Reviewer 2: Yeah, strike that one. That one should be used on the created warning.

Author: Done

### Usage

> The module exports two builder functions for creating warnings. A warning has a `emit` method.

Reviewer 2: ```suggestion
The module exports two builder functions for creating warnings. A warning has an `emit` method.
```

Reviewer: Actually forgot to remove this in #97 
```suggestion
The module exports two builder functions for creating warnings.
```

```js
const {
  createWarning,
  createDeprecation
} = require('process-warning')

> const warning = createWarning({
>   name: 'FastifyWarning',
>   code: 'FSTWRN001',
>   message: 'Hello %s',
>   unlimited: true
> })

Reviewer 2: I think we should avoid referencing Fastify in the examples:

1. Because this is a generic ecosystem module and is merely used by Fastify.
2. Because it can give the reader the impression that this module is only useful in Fastify applications.

```suggestion
const warning = createWarning({
  name: 'ExampleWarning',
  code: 'EXP_WRN_001',
  message: 'Hello %s',
  unlimited: true
})
```

>   once? Defaults to `false`.

Reviewer 2: We should flip this to `true`. It is more likely that a warning should be issued an infinite number of times than a finite number. Singular warnings will get lost in regular application logs.

Author: Logs cost money often, so I like that the explicit `unlimited` bool so devs know that the warning is being emitted more and more times

Reviewer 2: That is true. And I will concede that most users of this library are probably writing their own libraries for which they want users to be aware of deprecations, and they likely won't consider that detail. But I think that is also an argument for making the default `true`: it is the most common intention. People the warnings are meant to educate are unlikely to see a one off warning, and will then complain when the feature they were relying has been removed.

@fastify/core please cast a vote here:
👍 => `true`
👎 => `false`

##### `createDeprecation({code, message[, options]})`

This is a wrapper for `warning.create`. It is equivalent to invoking
`warning.create` with the `name` parameter set to "DeprecationWarning".

Deprecation warnings have extended support for the Node.js CLI options:
`--throw-deprecation`, `--no-deprecation`, and `--trace-deprecation`.

##### `warning.emit([, a [, b [, c]]])`

The utility also contains an `emit` function that you can use for emitting the
warnings you have previously created by passing their respective code.
A warning is guaranteed to be emitted at least once.

- `[, a [, b [, c]]]` (`any`, optional) - Parameters for string interpolation.

```js
const { createWarning } = require('process-warning')
const FST_ERROR_CODE = createWarning({ name: 'FastifyWarning', code: 'FST_ERROR_CODE', message: 'message' })

> FST_ERROR_CODE.emit()

Reviewer 2: ```suggestion
const { createWarning } = require('process-warning')
const EXP_ERROR_CODE = createWarning({ name: 'ExampleWarning', code: 'EXP_ERROR_CODE', message: 'message' })
EXP_ERROR_CODE.emit()
```

```

How to use an interpolated string:
```js
const { createWarning } = require('process-warning')
const FST_ERROR_CODE = createWarning({ name: 'FastifyWarning', code: 'FST_ERROR_CODE', message: 'Hello %s'})

> FST_ERROR_CODE.emit('world')

Reviewer 2: ```suggestion
const { createWarning } = require('process-warning')
const EXP_ERROR_CODE = createWarning({ name: 'ExampleWarning', code: 'EXP_ERROR_CODE', message: 'Hello %s'})
EXP_ERROR_CODE.emit('world')
```

```

The `warning` object has methods and properties for managing the warning's state. Useful for testing.
```js
const { createWarning } = require('process-warning')
const FST_ERROR_CODE = createWarning({ name: 'FastifyWarning', code: 'FST_ERROR_CODE', message: 'Hello %s'})
console.log(FST_ERROR_CODE.emitted) // false
FST_ERROR_CODE.emit('world')
console.log(FST_ERROR_CODE.emitted) // true

const FST_ERROR_CODE_2 = createWarning('FastifyWarning', 'FST_ERROR_CODE_2', 'Hello %s')
FST_ERROR_CODE_2.emitted = true

> FST_ERROR_CODE_2.emit('world') // will not be emitted

Reviewer 2: ```suggestion
const EXP_ERROR_CODE = createWarning({ name: 'ExampleWarning', code: 'EXP_ERROR_CODE', message: 'Hello %s'})
console.log(EXP_ERROR_CODE.emitted) // false
EXP_ERROR_CODE.emit('world')
console.log(EXP_ERROR_CODE.emitted) // true

const EXP_ERROR_CODE_2 = createWarning({ name: 'ExampleWarning', code: 'EXP_ERROR_CODE_2', message: 'Hello %s' })
EXP_ERROR_CODE_2.emitted = true
EXP_ERROR_CODE_2.emit('world') // will not be emitted
```

```

How to use an unlimited warning:
```js
const { createWarning } = require('process-warning')
const FST_ERROR_CODE = createWarning({ name: 'FastifyWarning', code: 'FST_ERROR_CODE', message: 'Hello %s', unlimited: true })
FST_ERROR_CODE.emit('world') // will be emitted

> FST_ERROR_CODE.emit('world') // will be emitted again

Reviewer 2: ```suggestion
const { createWarning } = require('process-warning')
const EXP_ERROR_CODE = createWarning({ name: 'ExampleWarning', code: 'EXP_ERROR_CODE', message: 'Hello %s', unlimited: true })
EXP_ERROR_CODE.emit('world') // will be emitted
EXP_ERROR_CODE.emit('world') // will be emitted again
```
